// safevalues/dist/cjs/dom/elements/element.js
"use strict";
/**
 * @license
 * Copyright Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setElementAttribute = exports.setElementPrefixedAttribute = exports.buildPrefixedAttributeSetter = exports.elementInsertAdjacentHtml = exports.setElementOuterHtml = exports.setElementInnerHtml = void 0;
/**
 * @fileoverview This contains safe wrappers for properties that aren't specific
 * to one kind of HTMLElement (like innerHTML), plus other setters and functions
 * that are not tied to elements (like location.href or Worker constructor).
 */
require("../../environment/dev.js");
const attribute_impl_js_1 = require("../../internals/attribute_impl.js");
const html_impl_js_1 = require("../../internals/html_impl.js");
const anchor_js_1 = require("./anchor.js");
const area_js_1 = require("./area.js");
const base_js_1 = require("./base.js");
const button_js_1 = require("./button.js");
const embed_js_1 = require("./embed.js");
const form_js_1 = require("./form.js");
const iframe_js_1 = require("./iframe.js");
const input_js_1 = require("./input.js");
const object_js_1 = require("./object.js");
const script_js_1 = require("./script.js");
/**
 * Safely set {@link Element.innerHTML} on a given ShadowRoot or Element which
 * may not be a `<script>` element or a `<style>` element.
 */
function setElementInnerHtml(elOrRoot, v) {
    if (isElement(elOrRoot)) {
        throwIfScriptOrStyle(elOrRoot);
    }
    elOrRoot.innerHTML = (0, html_impl_js_1.unwrapHtml)(v);
}
exports.setElementInnerHtml = setElementInnerHtml;
/**
 * Safely set {@link Element.outerHTML} for the given Element.
 */
function setElementOuterHtml(e, v) {
    const parent = e.parentElement;
    if (parent !== null) {
        throwIfScriptOrStyle(parent);
    }
    e.outerHTML = (0, html_impl_js_1.unwrapHtml)(v);
}
exports.setElementOuterHtml = setElementOuterHtml;
/**
 * Safely call {@link Element.insertAdjacentHTML} for the given Element.
 */
function elementInsertAdjacentHtml(element, position, v) {
    const tagContext = position === 'beforebegin' || position === 'afterend'
        ? element.parentElement
        : element;
    if (tagContext !== null) {
        throwIfScriptOrStyle(tagContext);
    }
    element.insertAdjacentHTML(position, (0, html_impl_js_1.unwrapHtml)(v));
}
exports.elementInsertAdjacentHtml = elementInsertAdjacentHtml;
/**
 * Given a set of known-to-be-safe prefixes (e.g., "data-", "aria-", "js"),
 * return a setter function that allows you to set attributes on an element,
 * as long as the names of the attributes to be set has one of the prefixes.
 *
 * The returned setter ensures that setting any dangerous attribute, e.g.,
 * "src", "href" will cause an exception. This is intended to be used as the
 * safe alterantive of `Element#setAttribute`, when applications need to set
 * attributes that do not have security implications and do not have a
 * corresponding DOM property.
 */
function buildPrefixedAttributeSetter(prefix, ...otherPrefixes) {
    const prefixes = [prefix, ...otherPrefixes];
    return (e, attr, value) => {
        setElementPrefixedAttribute(prefixes, e, attr, value);
    };
}
exports.buildPrefixedAttributeSetter = buildPrefixedAttributeSetter;
/**
 * A safe alternative to Element#setAttribute. The function takes a list of
 * `SafeAttributePrefix`, making developer intention explicit. The attribute
 * to be set must has one of the safe prefixes, otherwise the function throws
 * an Error.
 */
function setElementPrefixedAttribute(attrPrefixes, e, attr, value) {
    if (attrPrefixes.length === 0) {
        let message = '';
        if (process.env.NODE_ENV !== 'production') {
            message = 'No prefixes are provided';
        }
        throw new Error(message);
    }
    const prefixes = attrPrefixes.map((s) => (0, attribute_impl_js_1.unwrapAttributePrefix)(s));
    const attrLower = attr.toLowerCase();
    if (prefixes.every((p) => attrLower.indexOf(p) !== 0)) {
        throw new Error(`Attribute "${attr}" does not match any of the allowed prefixes.`);
    }
    e.setAttribute(attr, value);
}
exports.setElementPrefixedAttribute = setElementPrefixedAttribute;
function throwIfScriptOrStyle(element) {
    let message = '';
    const tagName = element.tagName;
    if (/^(script|style)$/i.test(tagName)) {
        if (process.env.NODE_ENV !== 'production') {
            if (tagName.toLowerCase() === 'script') {
                message = 'Use setScriptTextContent with a SafeScript.';
            }
            else {
                message = 'Use setStyleTextContent with a SafeStyleSheet.';
            }
        }
        throw new Error(message);
    }
}
function isElement(elOrRoot) {
    return elOrRoot.nodeType === 1; // Node.ELEMENT_NODE
}
/**
 * A safe alternative to Element#setAttribute.
 *
 * The function has essentially the same signature as `Element.setAttribute`,
 * but requires a safe type (or sanitizes the value) when used with a security
 * sensitive attribute. It does this by forwarding the call to the
 * element-specific setters within `safevalues/dom`.
 *
 * Note that this function doesn't currently support elements outside of the
 * html namespace & might throw if used with the wrong type of element or
 * attribute value
 *
 * If code size is a concern, consider using `setPrefixedAttribute`, or the
 * element-specific setters.
 *
 * The security sensitive element/attributes pairs are the following:
 *   - anchor#href -> forwarded to `setAnchorHref`
 *   - area#href -> forwarded to `setAreaHref`
 *   - base#href -> forwarded to `setBaseHref`
 *   - button#formaction -> forwarded to `setButtonFormaction`
 *   - embed#src -> forwarded to `setEmbedSrc`
 *   - form#action -> forwarded to `setFormAction`
 *   - iframe#src -> forwarded to `setIframeSrc`
 *   - iframe#srcdoc -> forwarded to `setIframeSrcdoc`
 *   - iframe#sandbox -> rejected, use `setIframeSrcWithIntent` or
 *       `setIframeSrcdocWithIntent` instead
 *   - input#formaction -> forwarded to `setInputFormaction`
 *   - link#href -> rejected, use `setLinkHrefAndRel` instead
 *   - link#rel -> rejected, use `setLinkHrefAndRel` instead
 *   - object#data -> forwarded to `setObjectData`
 *   - script#src -> forwarded to `setScriptSrc`
 *   - global attributes:
 *   - target -> forwarded to `el.setAttribute`
 *   - cite -> forwarded to `el.setAttribute`
 *   - poster -> forwarded to `el.setAttribute`
 *   - srcset -> forwarded to `el.setAttribute`
 *   - src -> forwarded to `el.setAttribute`
 *   - href -> forwarded to `el.setAttribute`
 *   - any attribute starting with `on` -> rejected
 *
 * Every other attribute is set as is using `element.setAttribute`
 */
function setElementAttribute(el, attr, value) {
    if (el.namespaceURI !== 'http://www.w3.org/1999/xhtml') {
        throw new Error(`Cannot set attribute '${attr}' on '${el.tagName}'.` +
            `Element is not in the HTML namespace`);
    }
    attr = attr.toLowerCase();
    const key = `${el.tagName} ${attr}`;
    switch (key) {
        case 'A href':
            (0, anchor_js_1.setAnchorHref)(el, value);
            return;
        case 'AREA href':
            (0, area_js_1.setAreaHref)(el, value);
            return;
        case 'BASE href':
            (0, base_js_1.setBaseHref)(el, value);
            return;
        case 'BUTTON formaction':
            (0, button_js_1.setButtonFormaction)(el, value);
            return;
        case 'EMBED src':
            (0, embed_js_1.setEmbedSrc)(el, value);
            return;
        case 'FORM action':
            (0, form_js_1.setFormAction)(el, value);
            return;
        case 'IFRAME src':
            (0, iframe_js_1.setIframeSrc)(el, value);
            return;
        case 'IFRAME srcdoc':
            (0, iframe_js_1.setIframeSrcdoc)(el, value);
            return;
        case 'IFRAME sandbox':
            throw new Error("Can't set 'sandbox' on iframe tags. " +
                'Use setIframeSrcWithIntent or setIframeSrcdocWithIntent instead');
        case 'INPUT formaction':
            (0, input_js_1.setInputFormaction)(el, value);
            return;
        case 'LINK href':
            throw new Error("Can't set 'href' attribute on link tags. " +
                'Use setLinkHrefAndRel instead');
        case 'LINK rel':
            throw new Error("Can't set 'rel' attribute on link tags. " +
                'Use setLinkHrefAndRel instead');
        case 'OBJECT data':
            (0, object_js_1.setObjectData)(el, value);
            return;
        case 'SCRIPT src':
            (0, script_js_1.setScriptSrc)(el, value);
            return;
        default:
            if (/^on./.test(attr)) {
                throw new Error(`Attribute "${attr}" looks like an event handler attribute. ` +
                    `Please use a safe alternative like addEventListener instead.`);
            }
            el.setAttribute(attr, value);
    }
}
exports.setElementAttribute = setElementAttribute;
