// safevalues/dist/cjs/internals/script_impl.js
"use strict";
/**
 * @license
 * Copyright Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.unwrapScript = exports.isScript = exports.EMPTY_SCRIPT = exports.createScriptInternal = exports.SafeScript = void 0;
/** @fileoverview Internal implementations of SafeScript. */
const trusted_types_js_1 = require("./trusted_types.js");
require("../environment/dev.js");
const pure_js_1 = require("./pure.js");
const secrets_js_1 = require("./secrets.js");
/**
 * JavaScript code that is safe to evaluate and use as the content of an HTML
 * script element.
 *
 * @final
 */
class SafeScript {
    constructor(token, value) {
        if (process.env.NODE_ENV !== 'production') {
            (0, secrets_js_1.ensureTokenIsValid)(token);
        }
        this.privateDoNotAccessOrElseWrappedScript = value;
    }
    toString() {
        // String coercion minimizes code size.
        // tslint:disable-next-line:restrict-plus-operands
        return this.privateDoNotAccessOrElseWrappedScript + '';
    }
}
exports.SafeScript = SafeScript;
const ScriptImpl = SafeScript;
function constructScript(value) {
    return new ScriptImpl(secrets_js_1.secretToken, value);
}
/**
 * Builds a new `SafeScript` from the given string, without enforcing
 * safety guarantees. It may cause side effects by creating a Trusted Types
 * policy. This shouldn't be exposed to application developers, and must only be
 * used as a step towards safe builders or safe constants.
 */
function createScriptInternal(value) {
    // Inlining this variable can cause large codesize increases when it is a
    // large constant string. See sizetests/examples/constants for an example.
    /** @noinline */
    const noinlineValue = value;
    const policy = (0, trusted_types_js_1.getPolicy)();
    return constructScript(policy ? policy.createScript(noinlineValue) : noinlineValue);
}
exports.createScriptInternal = createScriptInternal;
/**
 * An empty `SafeScript` constant.
 * Unlike the functions above, using this will not create a policy.
 */
exports.EMPTY_SCRIPT = (0, pure_js_1.pure)(() => constructScript(trusted_types_js_1.trustedTypes ? trusted_types_js_1.trustedTypes.emptyScript : ''));
/** Checks if the given value is a `SafeScript` instance */
function isScript(value) {
    return value instanceof SafeScript;
}
exports.isScript = isScript;
/**
 * Returns the value of the passed `SafeScript` object while ensuring it
 * has the correct type.
 * Using this function directly is not common. Safe types are not meant to be
 * unwrapped, but rather passed to other APIs that consume them, like the DOM
 * wrappers in safevalues/dom.
 *
 * Returns a native `TrustedScript` instance typed as {toString(): string} or a string if Trusted Types are disabled.
 */
function unwrapScript(value) {
    if (isScript(value)) {
        return value
            .privateDoNotAccessOrElseWrappedScript;
    }
    else {
        let message = '';
        if (process.env.NODE_ENV !== 'production') {
            message = 'Unexpected type when unwrapping SafeScript';
        }
        throw new Error(message);
    }
}
exports.unwrapScript = unwrapScript;
